src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+import math
+import logging
+from collections import defaultdict
+
+from src.providers.keys.typings import LidoKey
+from src.utils.helpers import get_first_non_missed_slot
+from src.web3_extentions import LidoValidator
 from src.web3_extentions.typings import Web3
 
+from src.modules.accounting.typings import Gwei
+from src.modules.submodules.consensus import FrameConfig, ChainConfig
+from src.providers.consensus.typings import ValidatorStatus, Validator
+from src.typings import BlockStamp, SlotNumber, EpochNumber
+
+# Constants from consensus spec
+MIN_VALIDATOR_WITHDRAWABILITY_DELAY = 256
+EPOCHS_PER_SLASHINGS_VECTOR = 8192
+PROPORTIONAL_SLASHING_MULTIPLIER_BELLATRIX = 3
+EFFECTIVE_BALANCE_INCREMENT = 2 ** 0 * 10 ** 9
+MIN_DEPOSIT_AMOUNT = 32 * 10 ** 9
+
+FIRST_DEPOSIT_EVENT_LOOKUP_MAXIMUM_BLOCKS_DISTANCE = 225 * 7
+
+NORMALIZED_CL_PER_EPOCH = 64
+MISTAKE_RANGE = 0.1
+NEAREST_EPOCH_DISTANCE = 4
+FAR_EPOCH_DISTANCE = 25
+
+logger = logging.getLogger(__name__)
+
 
 class BunkerService:
-    def __init__(self, w3: Web3):
+
+    def __init__(
+        self,
+        w3: Web3,
+        frame_config: FrameConfig,
+        chain_config: ChainConfig
+    ):
         self.w3 = w3
+        self.f_conf = frame_config
+        self.c_conf = chain_config
+        # Will be filled in `is_bunker_mode` call
+        self.last_report_ref_slot = None
+        self.all_validators: list[Validator] = []
+        self.lido_keys: list[LidoKey] = []
+        self.lido_validators: list[LidoValidator] = []
+
+    def is_bunker_mode(self, blockstamp: BlockStamp) -> bool:
+        self.all_validators, self.lido_keys, self.lido_validators = self._get_lido_validators_with_others(blockstamp)
+        self.last_report_ref_slot = self.w3.lido_contracts.accounting_oracle.functions.getLastProcessingRefSlot().call(
+            block_identifier=blockstamp.block_hash
+        )
+        logger.info({"msg": f"Validators - all: {len(self.all_validators)} lido: {len(self.lido_validators)}"})
+        logger.info({"msg": "Checking bunker mode"})
+        frame_cl_rebase = self._get_cl_rebase_for_frame(blockstamp)
+        if frame_cl_rebase < 0:
+            logger.info({"msg": "Bunker ON. CL rebase is negative"})
+            return True
+        if self._is_high_midterm_slashing_penalty(blockstamp, frame_cl_rebase):
+            logger.info({"msg": "Bunker ON. High midterm slashing penalty"})
+            return True
+        if self._is_abnormal_cl_rebase(blockstamp, frame_cl_rebase):
+            logger.info({"msg": "Bunker ON. Abnormal CL rebase"})
+            return True
+
```

Comment:
I would suggest adding `return False` at the end of the method

Reply:
done